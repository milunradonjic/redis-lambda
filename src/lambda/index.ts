import { createClient } from 'redis';

type RedisClient = ReturnType<typeof createClient> & { connected?: boolean };
type Event = Record<string, any>;
type Response = { statusCode: number; body: string }

export const client: RedisClient = createClient({
  url: process.env.CACHE_URL,
});


/**
 * Lambda function handler
 *
 * This function processes the incoming event object, retrieves the key
 * from the event object (either from queryStringParameters or the event object itself),
 * and performs the following steps:
 * 1. Attempts to get the value associated with the provided key from Redis.
 * 2. If the value is not found in Redis:
 *    - Stores the value from the request (or an empty string if no value is provided) in Redis with the given key.
 *    - Sets a TTL of 1 hour for the key-value pair in Redis.
 * 3. Returns the key and value as part of the response object.
 *
 * If the key parameter is not provided, the function returns a 400 Bad Request response with an error message.
 * If an error occurs while processing the request, the function logs the error and returns a 500 Internal Server Error response.
 *
 * @param {Event} event - The incoming event object
 * @returns {Promise<Response>} The response object
 */
export async function handler(event: Event): Promise<Response> {
  console.debug('Event', { event });

  if (!client.connected) {
    await client.connect();
    client.connected = true;
  }
  // Retrieve the 'key' value from either the queryStringParameters or the event object itself.
  // We also set 'valueFromRequest' to the same value as 'key', treating the key as the default value.
  const key = event.queryStringParameters?.key || event?.key;
  const valueFromRequest = key;

  if (!key) {
    console.error('Key parameter is missing');
    return {
      statusCode: 400,
      body: JSON.stringify({ message: 'Key parameter is missing' }),
    };
  }

  try {
    console.debug('Get by key', key);
    let value = await client.get(key);
    if (value === null) {
      console.debug('Value not found. Setting it.');
      value = valueFromRequest;
      await client.set(key, valueFromRequest);
      await client.expire(key, 3600);
    }

    return {
      statusCode: 200,
      body: JSON.stringify({ key, value }),
    };
  } catch (error) {
    console.error(error);
    return {
      statusCode: 500,
      body: JSON.stringify({ message: 'Internal Server Error' }),
    };
  }
};
