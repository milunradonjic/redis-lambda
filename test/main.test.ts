import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { LambdaStack } from '../src/stacks/lambda.stack';
import { RedisStack } from '../src/stacks/redis.stack';

test('Snapshot', () => {
  const app = new App();
  const redisStack = new RedisStack(app, 'testRedis');
  const lambdaStack = new LambdaStack(app, 'testLambda', {
    redisCache: redisStack.redisCache,
    redisSG: redisStack.redisSecurityGroup,
    vpc: redisStack.vpc,
  });

  const redisTemplate = Template.fromStack(redisStack);
  const lambdaTemplate = Template.fromStack(lambdaStack);
  expect(redisTemplate.toJSON()).toMatchSnapshot();
  expect(lambdaTemplate.toJSON()).toMatchSnapshot();
});