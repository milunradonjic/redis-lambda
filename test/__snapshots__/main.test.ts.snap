// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Outputs": {
    "ExportsOutputFnGetAttRedisStackredisCacheRedisEndpointAddress14A525E0": {
      "Export": {
        "Name": "testRedis:ExportsOutputFnGetAttRedisStackredisCacheRedisEndpointAddress14A525E0",
      },
      "Value": {
        "Fn::GetAtt": [
          "RedisStackredisCache",
          "RedisEndpoint.Address",
        ],
      },
    },
    "ExportsOutputFnGetAttRedisStackredisCacheRedisEndpointPort5F3953BD": {
      "Export": {
        "Name": "testRedis:ExportsOutputFnGetAttRedisStackredisCacheRedisEndpointPort5F3953BD",
      },
      "Value": {
        "Fn::GetAtt": [
          "RedisStackredisCache",
          "RedisEndpoint.Port",
        ],
      },
    },
    "ExportsOutputFnGetAttRedisStackredisSecurityGroup44892807GroupId9FFA6681": {
      "Export": {
        "Name": "testRedis:ExportsOutputFnGetAttRedisStackredisSecurityGroup44892807GroupId9FFA6681",
      },
      "Value": {
        "Fn::GetAtt": [
          "RedisStackredisSecurityGroup44892807",
          "GroupId",
        ],
      },
    },
    "ExportsOutputRefRedisStackVpc876ADC10AB97EB7C": {
      "Export": {
        "Name": "testRedis:ExportsOutputRefRedisStackVpc876ADC10AB97EB7C",
      },
      "Value": {
        "Ref": "RedisStackVpc876ADC10",
      },
    },
    "ExportsOutputRefRedisStackVpcRedisStackPrivateSubnetSubnet1Subnet987203F7A2D5756B": {
      "Export": {
        "Name": "testRedis:ExportsOutputRefRedisStackVpcRedisStackPrivateSubnetSubnet1Subnet987203F7A2D5756B",
      },
      "Value": {
        "Ref": "RedisStackVpcRedisStackPrivateSubnetSubnet1Subnet987203F7",
      },
    },
    "RedisStackCacheEndpointUrl": {
      "Value": {
        "Fn::GetAtt": [
          "RedisStackredisCache",
          "RedisEndpoint.Address",
        ],
      },
    },
    "RedisStackCachePort": {
      "Value": {
        "Fn::GetAtt": [
          "RedisStackredisCache",
          "RedisEndpoint.Port",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "RedisStackVpc876ADC10": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "testRedis/RedisStackVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "RedisStackVpcIGW4E405A0A": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "testRedis/RedisStackVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "RedisStackVpcRedisStackPrivateSubnetSubnet1RouteTableAssociation7BD56A5D": {
      "Properties": {
        "RouteTableId": {
          "Ref": "RedisStackVpcRedisStackPrivateSubnetSubnet1RouteTableCEA88F55",
        },
        "SubnetId": {
          "Ref": "RedisStackVpcRedisStackPrivateSubnetSubnet1Subnet987203F7",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "RedisStackVpcRedisStackPrivateSubnetSubnet1RouteTableCEA88F55": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "testRedis/RedisStackVpc/RedisStackPrivateSubnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "RedisStackVpc876ADC10",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "RedisStackVpcRedisStackPrivateSubnetSubnet1Subnet987203F7": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/17",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "RedisStackPrivateSubnet",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          {
            "Key": "Name",
            "Value": "testRedis/RedisStackVpc/RedisStackPrivateSubnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "RedisStackVpc876ADC10",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "RedisStackVpcRedisStackPublicSubnetSubnet1DefaultRoute52D59DC1": {
      "DependsOn": [
        "RedisStackVpcVPCGWC108A861",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "RedisStackVpcIGW4E405A0A",
        },
        "RouteTableId": {
          "Ref": "RedisStackVpcRedisStackPublicSubnetSubnet1RouteTable54239E15",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RedisStackVpcRedisStackPublicSubnetSubnet1EIPE8082221": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "testRedis/RedisStackVpc/RedisStackPublicSubnetSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "RedisStackVpcRedisStackPublicSubnetSubnet1NATGateway61413E3D": {
      "DependsOn": [
        "RedisStackVpcRedisStackPublicSubnetSubnet1DefaultRoute52D59DC1",
        "RedisStackVpcRedisStackPublicSubnetSubnet1RouteTableAssociation82F82CFB",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "RedisStackVpcRedisStackPublicSubnetSubnet1EIPE8082221",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "RedisStackVpcRedisStackPublicSubnetSubnet1Subnet3F6E7427",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "testRedis/RedisStackVpc/RedisStackPublicSubnetSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "RedisStackVpcRedisStackPublicSubnetSubnet1RouteTable54239E15": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "testRedis/RedisStackVpc/RedisStackPublicSubnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "RedisStackVpc876ADC10",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "RedisStackVpcRedisStackPublicSubnetSubnet1RouteTableAssociation82F82CFB": {
      "Properties": {
        "RouteTableId": {
          "Ref": "RedisStackVpcRedisStackPublicSubnetSubnet1RouteTable54239E15",
        },
        "SubnetId": {
          "Ref": "RedisStackVpcRedisStackPublicSubnetSubnet1Subnet3F6E7427",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "RedisStackVpcRedisStackPublicSubnetSubnet1Subnet3F6E7427": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/17",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "RedisStackPublicSubnet",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "testRedis/RedisStackVpc/RedisStackPublicSubnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "RedisStackVpc876ADC10",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "RedisStackVpcVPCGWC108A861": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "RedisStackVpcIGW4E405A0A",
        },
        "VpcId": {
          "Ref": "RedisStackVpc876ADC10",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "RedisStackredisCache": {
      "DependsOn": [
        "RedisStackredisSubnetGroup",
      ],
      "Properties": {
        "CacheNodeType": "cache.t3.micro",
        "CacheSubnetGroupName": {
          "Ref": "RedisStackredisSubnetGroup",
        },
        "ClusterName": "GT-Dev-Cluster",
        "Engine": "redis",
        "EngineVersion": "6.2",
        "NumCacheNodes": 1,
        "PreferredMaintenanceWindow": "fri:00:30-fri:01:30",
        "VpcSecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "RedisStackredisSecurityGroup44892807",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::ElastiCache::CacheCluster",
    },
    "RedisStackredisSecurityGroup44892807": {
      "Properties": {
        "GroupDescription": "Security group for the redis cluster",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": {
          "Ref": "RedisStackVpc876ADC10",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "RedisStackredisSubnetGroup": {
      "Properties": {
        "CacheSubnetGroupName": "GT-Redis-Subnet-Group",
        "Description": "Subnet group for the redis cluster",
        "SubnetIds": [
          {
            "Ref": "RedisStackVpcRedisStackPublicSubnetSubnet1Subnet3F6E7427",
          },
        ],
      },
      "Type": "AWS::ElastiCache::SubnetGroup",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Snapshot 2`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LambdaStacklambdaFn9080B89D": {
      "DependsOn": [
        "LambdaStacklambdaRole27DCF012",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "4559f03dc93637ac3872d7133eff9a0d51f83089921c1d6f1730ba7e44770d30.zip",
        },
        "Environment": {
          "Variables": {
            "CACHE_URL": {
              "Fn::Join": [
                "",
                [
                  "redis://",
                  {
                    "Fn::ImportValue": "testRedis:ExportsOutputFnGetAttRedisStackredisCacheRedisEndpointAddress14A525E0",
                  },
                  ":",
                  {
                    "Fn::ImportValue": "testRedis:ExportsOutputFnGetAttRedisStackredisCacheRedisEndpointPort5F3953BD",
                  },
                ],
              ],
            },
          },
        },
        "FunctionName": "redis-lambdaFn",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LambdaStacklambdaRole27DCF012",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 200,
        "VpcConfig": {
          "SecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "LambdaStacklambdaSG14DCD604",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": [
            {
              "Fn::ImportValue": "testRedis:ExportsOutputRefRedisStackVpcRedisStackPrivateSubnetSubnet1Subnet987203F7A2D5756B",
            },
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaStacklambdaRole27DCF012": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonElastiCacheFullAccess",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaENIManagementAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaStacklambdaSG14DCD604": {
      "Properties": {
        "GroupDescription": "testLambda/LambdaStacklambdaSG",
        "GroupName": "redis-lambdaFn Security Group",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": {
          "Fn::ImportValue": "testRedis:ExportsOutputRefRedisStackVpc876ADC10AB97EB7C",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "LambdaStacklambdaSGtestRedisRedisStackredisSecurityGroup8389985A6379toAE1B3963": {
      "Properties": {
        "Description": "Allow this lambda function connect to the redis cache",
        "FromPort": 6379,
        "GroupId": {
          "Fn::ImportValue": "testRedis:ExportsOutputFnGetAttRedisStackredisSecurityGroup44892807GroupId9FFA6681",
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "LambdaStacklambdaSG14DCD604",
            "GroupId",
          ],
        },
        "ToPort": 6379,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
